## Food/Diet Recommender System

#### Group04 ##

**Tutorial section: TT5L**

| | Student ID     | Student Name                |
|--:|:-------------|:---------------------------:|
| 1 | 1201103207   | Melody Koh Si Jie           | 
| 2 | 1211103427   | Law Chin Keat               |
| 3 | 1211102976   | Lee Le Xuan                 | 
| 4 | 1211200755   | Wendy Tan Wei Wei           | 

### 1.0 Import libraries

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
df = pd.read_csv('data/RAW_recipes.csv')

In [3]:
df.columns

Index(['name', 'id', 'minutes', 'contributor_id', 'submitted', 'tags',
       'nutrition', 'n_steps', 'steps', 'description', 'ingredients',
       'n_ingredients'],
      dtype='object')

In [4]:
df.dtypes

name              object
id                 int64
minutes            int64
contributor_id     int64
submitted         object
tags              object
nutrition         object
n_steps            int64
steps             object
description       object
ingredients       object
n_ingredients      int64
dtype: object

In [5]:
df.shape

(231637, 12)

In [6]:
df

name      id  minutes  \
0         arriba   baked winter squash mexican style  137739       55   
1                   a bit different  breakfast pizza   31490       30   
2                          all in the kitchen  chili  112140      130   
3                                 alouette  potatoes   59389       45   
4                 amish  tomato ketchup  for canning   44061      190   
...                                              ...     ...      ...   
231632                                   zydeco soup  486161       60   
231633                              zydeco spice mix  493372        5   
231634                     zydeco ya ya deviled eggs  308080       40   
231635        cookies by design   cookies on a stick  298512       29   
231636  cookies by design   sugar shortbread cookies  298509       20   

        contributor_id   submitted  \
0                47892  2005-09-16   
1                26278  2002-06-17   
2               196586  2005-02-25   
3                68585  2003-04-14   
4                41706  2002-10-25   
...                ...         ...   
231632          227978  2012-08-29   
231633         1500678  2013-01-09   
231634           37779  2008-06-07   
231635          506822  2008-04-15   
231636          506822  2008-04-15   

                                                     tags  \
0       ['60-minutes-or-less', 'time-to-make', 'course...   
1       ['30-minutes-or-less', 'time-to-make', 'course...   
2       ['time-to-make', 'course', 'preparation', 'mai...   
3       ['60-minutes-or-less', 'time-to-make', 'course...   
4       ['weeknight', 'time-to-make', 'course', 'main-...   
...                                                   ...   
231632  ['ham', '60-minutes-or-less', 'time-to-make', ...   
231633  ['15-minutes-or-less', 'time-to-make', 'course...   
231634  ['60-minutes-or-less', 'time-to-make', 'course...   
231635  ['30-minutes-or-less', 'time-to-make', 'course...   
231636  ['30-minutes-or-less', 'time-to-make', 'course...   

                                          nutrition  n_steps  \
0             [51.5, 0.0, 13.0, 0.0, 2.0, 0.0, 4.0]       11   
1         [173.4, 18.0, 0.0, 17.0, 22.0, 35.0, 1.0]        9   
2        [269.8, 22.0, 32.0, 48.0, 39.0, 27.0, 5.0]        6   
3         [368.1, 17.0, 10.0, 2.0, 14.0, 8.0, 20.0]       11   
4         [352.9, 1.0, 337.0, 23.0, 3.0, 0.0, 28.0]        5   
...                                             ...      ...   
231632  [415.2, 26.0, 34.0, 26.0, 44.0, 21.0, 15.0]        7   
231633        [14.8, 0.0, 2.0, 58.0, 1.0, 0.0, 1.0]        1   
231634         [59.2, 6.0, 2.0, 3.0, 6.0, 5.0, 0.0]        7   
231635    [188.0, 11.0, 57.0, 11.0, 7.0, 21.0, 9.0]        9   
231636     [174.9, 14.0, 33.0, 4.0, 4.0, 11.0, 6.0]        5   

                                                    steps  \
0       ['make a choice and proceed with recipe', 'dep...   
1       ['preheat oven to 425 degrees f', 'press dough...   
2       ['brown ground beef in large pot', 'add choppe...   
3       ['place potatoes in a large pot of lightly sal...   
4       ['mix all ingredients& boil for 2 1 / 2 hours ...   
...                                                   ...   
231632  ['heat oil in a 4-quart dutch oven', 'add cele...   
231633        ['mix all ingredients together thoroughly']   
231634  ['in a bowl , combine the mashed yolks and may...   
231635  ['place melted butter in a large mixing bowl a...   
231636  ['whip sugar and shortening in a large bowl , ...   

                                              description  \
0       autumn is my favorite time of year to cook! th...   
1       this recipe calls for the crust to be prebaked...   
2       this modified version of 'mom's' chili was a h...   
3       this is a super easy, great tasting, make ahea...   
4       my dh's amish mother raised him on this recipe...   
...                                                   ...   
231632  this is a delicious soup that i originally f

### 2.0 Missing values handing

In [7]:
df.isnull().sum().sum()

4980

In [8]:
4980/231637*100

2.149915600702824

there are 4980 rows which has at least one null value, it seems to be a lot, but 4980/231637 = 0.02 * 100 which is only 2.15%. Now let's try to look at the missing value and analyse them

In [9]:
df.isnull().sum()

name                 1
id                   0
minutes              0
contributor_id       0
submitted            0
tags                 0
nutrition            0
n_steps              0
steps                0
description       4979
ingredients          0
n_ingredients        0
dtype: int64

and we can see that there is one row of name is missing and all the remaining missing values are description. We first handle on column name

In [10]:
# filter on column name which row is having null value
name_null = df['name'].isnull()
df[name_null]

name      id  minutes  contributor_id   submitted  \
721  NaN  368257       10          779451  2009-04-27   

                                                  tags  \
721  ['15-minutes-or-less', 'time-to-make', 'course...   

                                         nutrition  n_steps  \
721  [1596.2, 249.0, 155.0, 0.0, 2.0, 112.0, 14.0]        6   

                                                 steps    description  \
721  ['in a bowl , combine ingredients except for o...  -------------   

                                           ingredients  n_ingredients  
721  ['lemon', 'honey', 'horseradish mustard', 'gar...             10

we noticed that row 721 is missing with the name, but based on the ingredients and steps, and know that it is a salad dressing, and we can fill it manually

In [11]:
df.fillna({'name': 'salad dressing'}, inplace=True)
df.loc[721]

name                                                 salad dressing
id                                                           368257
minutes                                                          10
contributor_id                                               779451
submitted                                                2009-04-27
tags              ['15-minutes-or-less', 'time-to-make', 'course...
nutrition             [1596.2, 249.0, 155.0, 0.0, 2.0, 112.0, 14.0]
n_steps                                                           6
steps             ['in a bowl , combine ingredients except for o...
description                                           -------------
ingredients       ['lemon', 'honey', 'horseradish mustard', 'gar...
n_ingredients                                                    10
Name: 721, dtype: object

next, we need to handle the having most missing value column, description. From the description of every column in Kaggle, we know that column description is just the user-provided description which is an unnecessary for our model building, we can simply drop it

In [12]:
df.drop(['description'], axis=1, inplace=True)

In [13]:
# every missing values is now handled successfully
df.isnull().sum()

name              0
id                0
minutes           0
contributor_id    0
submitted         0
tags              0
nutrition         0
n_steps           0
steps             0
ingredients       0
n_ingredients     0
dtype: int64

### 2.1 Drop unnecessary column

we have dropped column ingredients before, lets analyse on other columns and we see and drop if there is any other unnecessary column for our model building later

In [14]:
df.columns

Index(['name', 'id', 'minutes', 'contributor_id', 'submitted', 'tags',
       'nutrition', 'n_steps', 'steps', 'ingredients', 'n_ingredients'],
      dtype='object')

1. we notice that contributor_id is simply just the user id who submitted the recipe which is totally useless for our model building
2. column submitted is the date recipe was submitted which is also unnecessary undoubtedly

In [15]:
df.drop(['contributor_id', 'submitted'], axis=1, inplace=True)

In [16]:
df

name      id  minutes  \
0         arriba   baked winter squash mexican style  137739       55   
1                   a bit different  breakfast pizza   31490       30   
2                          all in the kitchen  chili  112140      130   
3                                 alouette  potatoes   59389       45   
4                 amish  tomato ketchup  for canning   44061      190   
...                                              ...     ...      ...   
231632                                   zydeco soup  486161       60   
231633                              zydeco spice mix  493372        5   
231634                     zydeco ya ya deviled eggs  308080       40   
231635        cookies by design   cookies on a stick  298512       29   
231636  cookies by design   sugar shortbread cookies  298509       20   

                                                     tags  \
0       ['60-minutes-or-less', 'time-to-make', 'course...   
1       ['30-minutes-or-less', 'time-to-make', 'course...   
2       ['time-to-make', 'course', 'preparation', 'mai...   
3       ['60-minutes-or-less', 'time-to-make', 'course...   
4       ['weeknight', 'time-to-make', 'course', 'main-...   
...                                                   ...   
231632  ['ham', '60-minutes-or-less', 'time-to-make', ...   
231633  ['15-minutes-or-less', 'time-to-make', 'course...   
231634  ['60-minutes-or-less', 'time-to-make', 'course...   
231635  ['30-minutes-or-less', 'time-to-make', 'course...   
231636  ['30-minutes-or-less', 'time-to-make', 'course...   

                                          nutrition  n_steps  \
0             [51.5, 0.0, 13.0, 0.0, 2.0, 0.0, 4.0]       11   
1         [173.4, 18.0, 0.0, 17.0, 22.0, 35.0, 1.0]        9   
2        [269.8, 22.0, 32.0, 48.0, 39.0, 27.0, 5.0]        6   
3         [368.1, 17.0, 10.0, 2.0, 14.0, 8.0, 20.0]       11   
4         [352.9, 1.0, 337.0, 23.0, 3.0, 0.0, 28.0]        5   
...                                             ...      ...   
231632  [415.2, 26.0, 34.0, 26.0, 44.0, 21.0, 15.0]        7   
231633        [14.8, 0.0, 2.0, 58.0, 1.0, 0.0, 1.0]        1   
231634         [59.2, 6.0, 2.0, 3.0, 6.0, 5.0, 0.0]        7   
231635    [188.0, 11.0, 57.0, 11.0, 7.0, 21.0, 9.0]        9   
231636     [174.9, 14.0, 33.0, 4.0, 4.0, 11.0, 6.0]        5   

                                                    steps  \
0       ['make a choice and proceed with recipe', 'dep...   
1       ['preheat oven to 425 degrees f', 'press dough...   
2       ['brown ground beef in large pot', 'add choppe...   
3       ['place potatoes in a large pot of lightly sal...   
4       ['mix all ingredients& boil for 2 1 / 2 hours ...   
...                                                   ...   
231632  ['heat oil in a 4-quart dutch oven', 'add cele...   
231633        ['mix all ingredients together thoroughly']   
231634  ['in a bowl , combine the mashed yolks and may...   
231635  ['place melted butter in a large mixing bowl a...   
231636  ['whip sugar and shortening in a large bowl , ...   

                                              ingredients  n_ingredients  
0       ['winter squash', 'mexican seasoning', 'mixed ...              7  
1       ['prepared pizza crust', 'sausage patty', 'egg...              6  
2       ['ground beef', 'yellow onions', 'diced tomato...             13  
3       ['spreadable cheese with garlic and herbs', 'n...             11  
4       ['tomato juice', 'apple cider vinegar', 'sugar...              8  
...                                                   ...            ...  
231632  ['celery', 'onion', 'green sweet pepper', 'gar...             22  
231633  ['paprika', 'salt', 'garlic powder', 'onion po...             13  
231634  ['hard-cooked eggs', 'mayonnaise', 'dijon must...              8  
231635  ['butter', 'eagle brand condensed milk', 'ligh...             10  
231636  ['granulated sugar', 'shortening', 'eggs', 'fl...              7  

[231637 rows x 9 columns]

### 2.2 Duplicated values handing

In [17]:
duplicate_data = df.duplicated()
df[duplicate_data]

Empty DataFrame
Columns: [name, id, minutes, tags, nutrition, n_steps, steps, ingredients, n_ingredients]
Index: []

we can see that there is no any duplicated data, which is great